finality-aleph/src/justification/requester.rs

```diff
@@ -202,6 +203,72 @@ where
     }
 
     pub fn finalized_number(&self) -> NumberFor<B> {
-        self.client.info().finalized_number
+        self.backend.blockchain().info().finalized_number
+    }
+
+    fn do_request(&mut self, hash: &<B as BlockT>::Hash, num: NumberFor<B>) {
+        debug!(target: "aleph-justification",
+               "We have block {:?} with hash {:?}. Requesting justification.", num, hash);
+        self.justification_request_scheduler.on_request_sent();
+        self.block_requester.request_justification(hash, num);
+    }
+
+    // We request justifications for all the children of last finalized block.
+    // Assuming that we request at the same pace that finalization is progressing, it ensures
+    // that we are up to date with finalization.
+    // We also request the child that it's on the same branch as top_wanted since a fork may happen
+    // somewhere in between them.
+    fn request_children(&mut self, info: &Info<B>) {
```

Comment:
There is one pretty annoying problem with this fix. 
It assumes that justifications for ALL blocks exist and are generally available via the substrate request.
What is true is that:
All "regular" justifications have been created at some point, but I don't think we have a strict guarantee that many nodes imported them (I think we know that at least one validator did, but not much beyond that). This would require going deep into some unpleasant details do understand it fully.
Also there are also "emergency finalizations" which complicate the picture quite a bit, because the expected way of using them would be to emergency finalize the last block of a session. So this way we would skip many.

Reply:
I think the only way to make sure this works, is to request, once upon a time, ALL descendants/ancestors (one of these, I always confuse them) of the last finalized block, corresponding to the last block in session.